lib/stages/integrateHFBData.cpp

```diff
@@ -0,0 +1,228 @@
+                  % (_num_frames_to_integrate * _samples_per_data_set));
+
+        // Get the no. of lost samples in this frame
+        total_lost_timesamples +=
+            get_lost_timesamples(compressed_lost_samples_buf, compress_buffer_ID);
+
+        // When all frames have been integrated output the result
+        if (get_fpga_seq_num(in_buf, in_buffer_ID)
+            >= fpga_seq_num_end_old + _samples_per_data_set) {
+
+            DEBUG("fpga_seq_num_end_old: {:d}, fpga_seq_num: {:d}", fpga_seq_num_end_old,
+                  fpga_seq_num);
+            // Increment the no. of lost frames if there are missing frames
+            total_lost_timesamples += fpga_seq_num_end_old - fpga_seq_num;
+
+            const float good_samples_frac =
+                (float)(total_timesamples - total_lost_timesamples) / total_timesamples;
+
+            // Normalise data
+            const float norm_frac = normaliseFrame(sum_data, in_buffer_ID);
+
+            // Only output integration if there are enough good samples
+            if (good_samples_frac >= _good_samples_threshold) {
+
+                // Create new metadata
+                allocate_new_metadata_object(out_buf, out_buffer_ID);
+
+                // Populate metadata
+                int64_t fpga_seq =
+                    fpga_seq_num_end_old - ((_num_frames_to_integrate - 1) * _samples_per_data_set);
+                set_fpga_seq_num_hfb(out_buf, out_buffer_ID, fpga_seq);
+
+                // Check if GPS time is set
+                if (!is_gps_global_time_set())
+                    set_gps_time_flag(out_buf, out_buffer_ID, 0);
+                else {
+                    set_gps_time_flag(out_buf, out_buffer_ID, 1);
+                    set_gps_time_hfb(out_buf, out_buffer_ID, compute_gps_time(fpga_seq));
+                }
+
+                set_norm_frac(out_buf, out_buffer_ID, norm_frac);
+                set_num_samples_integrated(out_buf, out_buffer_ID,
+                                           total_timesamples - total_lost_timesamples);
+                set_num_samples_expected(out_buf, out_buffer_ID, total_timesamples);
+
+                const stream_id_t stream_id =
+                    extract_stream_id(get_stream_id(in_buf, in_buffer_ID));
+                uint32_t freq_bin_num = bin_number_chime(&stream_id);
+                set_freq_bin_num(out_buf, out_buffer_ID, freq_bin_num);
+
+                mark_frame_full(out_buf, unique_name.c_str(), out_buffer_ID);
+
+                // Get a new output buffer
+                out_buffer_ID = (out_buffer_ID + 1) % out_buf->num_frames;
+                out_frame = wait_for_empty_frame(out_buf, unique_name.c_str(), out_buffer_ID);
+                if (out_frame == NULL)
+                    goto end_loop;
+
+                sum_data = (float*)out_buf->frames[out_buffer_ID];
+            } else
+                DEBUG("Integration discarded. Too many lost samples.");
```

Comment:
Style point, I'd like this to be a fully `{` open and `}` close.  I find it harder to read these one line if results when one part is not one line.